test/controllers/ConfirmSubmissionControllerSpec.scala

```diff
@@ -150,25 +152,112 @@ class ConfirmSubmissionControllerSpec extends BaseSpec with MockAuth with MockMa
         }
       }
     }
+
     authControllerChecks(TestConfirmSubmissionController.show("18AA"), fakeRequest)
   }
 
   "ConfirmSubmissionController .submit" when {
-    "user is authorised" should {
 
-      lazy val result: Future[Result] = TestConfirmSubmissionController.submit("18AA")(fakeRequest.withSession(
-        SessionKeys.mandationStatus -> MandationStatuses.nonMTDfB)
-      )
+    "user is authorised" when {
+
+      "valid session data exists" when {
+
+        val nineBoxData = Json.obj(
+          "box1" -> "10.01",
+          "box2" -> "10.02",
+          "box3" -> "10.03",
+          "box4" -> "10.04",
+          "box5" -> "10.05",
+          "box6" -> "10.06",
+          "box7" -> "10.07",
+          "box8" -> "10.08",
+          "box9" -> "10.09",
+          "flatRateScheme" -> false,
+          "start" -> "2019-01-02",
+          "end" -> "2019-01-02",
+          "due" -> "2019-01-02"
+        ).toString()
+
+        "submission to backend is successful" should {
+
+          lazy val result: Future[Result] = TestConfirmSubmissionController.submit("18AA")(fakeRequest.withSession(
+            "mtdNineBoxReturnData" -> nineBoxData,
+            SessionKeys.mandationStatus -> MandationStatuses.nonMTDfB
+          ))
+
+          "return 303" in {
+            mockAuthorise(mtdVatAuthorisedResponse)
+            mockVatReturnsService(Future.successful(Right(SubmissionSuccessModel("12345"))))
+            status(result) shouldBe Status.SEE_OTHER
+          }
+
+          s"redirect to ${controllers.routes.ConfirmationController.show()}" in {
+            redirectLocation(result) shouldBe Some(controllers.routes.ConfirmationController.show().url)
+          }
+
+          "remove form data from session" in {
+            result.session.get("mtdNineBoxReturnData") shouldBe None
+          }
+        }
+
+        "submission to backend is unsuccessful" should {
+
+          lazy val result: Future[Result] = TestConfirmSubmissionController.submit("18AA")(fakeRequest.withSession(
+            "mtdNineBoxReturnData" -> nineBoxData,
+            SessionKeys.mandationStatus -> MandationStatuses.nonMTDfB
+          ))
+
+          "return 500" in {
+            mockAuthorise(mtdVatAuthorisedResponse)
+            mockVatReturnsService(Future.successful(Left(UnexpectedJsonFormat)))
+            status(result) shouldBe Status.INTERNAL_SERVER_ERROR
+          }
+
+          "show ISE page" in {
+            Jsoup.parse(bodyOf(result)).title() shouldBe "Sorry, we are experiencing technical difficulties - 500"
+          }
+        }
+      }
+
+      "invalid session data exists" should {
+
+        val nineBoxData = Json.obj("box10" -> "why").toString()
+        lazy val result: Future[Result] = TestConfirmSubmissionController.submit("18AA")(fakeRequest.withSession(
+          "mtdNineBoxReturnData" -> nineBoxData,
```

Comment:
Can you use the SessionKey here please and the other examples within this file as well please?

Reply:
I prefer typing it in so its almost like a test for the session key identifier being correct too (similar to having a separate message lookup in tests)